Added `transform_to_data_extent()`, improved `transform()`

LucaMarconato opened the pull request:
# New API:
- Added utils function `transform_to_data_extent()`, useful to create a converted with downscaled images, where all the -transformations are applied and where 1 pixel = 1 physical unit.

# Improved robustness and tests of transformations
- Added `maintain_positioning` argument to`SpatialData.transform_to_coordinate_system()` and `SpatialData.transform_element_to_coordinate_system()`.
- Fixed bug: incorrect offset translation vector for raster data when calling the two above functions, `transform()` was not affected)
- Fixed bug: `transform()` on multiscale raster was returning the offset translation for the most low res scale, not scale0

Most importantly:
- `transform()` now requires the argument `to_coordinate_system` instead of `transformation` (unless `maintain_positioning` is `True`); in fact, non-explicit target coordinate systems may lead to unintuitive results when an element is aligned to multiple coordinate systems. This change is communicated to the user by an informative exception; in some cases backward compatibility could be ensured so only an informative warning is given.

Note:
The above bugs were not affecting the rest of the codebase because `query()`, `aggregate()`, ... make either use of `transform()` for non-raster data, either operate on the transformations directly, without transforming raster data.

melonora commented on src/spatialdata/_core/spatialdata.py:
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
+            transformed = transform(element, transformation=t, maintain_positioning=maintain_positioning)
+        else:
+            d = get_transformation(element, get_all=True)
+            assert isinstance(d, dict)
```
Idem here, assert is not required I believe.
```suggestion
```

LucaMarconato replied:
makes mypy complain

melonora commented on src/spatialdata/_core/spatialdata.py:
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
+                del d[target_coordinate_system]
+        if not maintain_positioning:
+            d = get_transformation(transformed, get_all=True)
+            assert isinstance(d, dict)
```
```suggestion
```

LucaMarconato replied:
makes mypy complain

melonora commented on src/spatialdata/_core/spatialdata.py:
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
+            # system), then the transformation to the target coordinate system is not needed # because the data is now
+            # already transformed; here we remove such transformation.
+            d = get_transformation(transformed, get_all=True)
+            assert isinstance(d, dict)
```
```suggestion
```

LucaMarconato replied:
makes mypy complain

melonora commented on src/spatialdata/_core/operations/rasterize.py:
```diff
@@ -448,7 +448,8 @@ def _(
     elif schema in (Image2DModel, Image3DModel):
-        kwargs = {}
+        kwargs = {"order": 0}
+        # kwargs = {}
```
```suggestion
```

melonora commented on src/spatialdata/_core/operations/transform.py:
```diff
@@ -132,124 +135,192 @@ def _prepend_transformation(
+    assert isinstance(to_prepend, BaseTransformation)
 
     d = get_transformation(element, get_all=True)
     assert isinstance(d, dict)
```
```suggestion
```

LucaMarconato replied:
makes mypy complain

melonora commented on src/spatialdata/_core/operations/transform.py:
```diff
@@ -132,124 +135,192 @@ def _prepend_transformation(
+
+    if not maintain_positioning:
+        d = get_transformation(data, get_all=True)
+        assert isinstance(d, dict)
```
```suggestion
```

LucaMarconato replied:
makes mypy complain

melonora reviewed (commented):
Very quick first pass:)

melonora commented on tests/core/operations/test_transform.py:
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
-# TODO: when the io for 3D images and 3D labels work, add those tests
-def test_transform_image_spatial_image(images: SpatialData):
-    sdata = SpatialData(images={k: v for k, v in images.images.items() if isinstance(v, SpatialImage)})
+def _postpone_transformation(
+    sdata: SpatialData, from_coordinate_system: str, to_coordinate_system: str, transformation: BaseTransformation
+):
+    for element in sdata._gen_elements_values():
+        d = get_transformation(element, get_all=True)
+        assert isinstance(d, dict)
+        assert len(d) == 1
+        t = d[from_coordinate_system]
+        sequence = Sequence([t, transformation])
+        set_transformation(element, sequence, to_coordinate_system)
```
Perhaps better to move this to a `_test_utils.py` in src/spatialdata.

LucaMarconato replied:
I would keep this function here, at least for the moment, because I think it's better for the user to do this operation manually. In this case we assume that len(d) == 1, but in more complex scenario you may have an intricate graph of transformations, and this function could compromise some alignments to other coordinate systems.

melonora commented on tests/core/operations/test_transform.py:
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
-    _unpad_rasters(padded)
-    # raise NotImplementedError("TODO: plot the images")
-    # raise NotImplementedError("TODO: compare the transformed images with the original ones")
+def _remove_coordinate_system(sdata: SpatialData, coordinate_system: str):
+    for element in sdata._gen_elements_values():
+        remove_transformation(element, coordinate_system)
```
Same as above

LucaMarconato replied:
moved to utils function `remove_transformations_to_coordinate_system()`

melonora commented on tests/core/operations/test_transform.py:
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
-    _unpad_rasters(padded)
-    # TODO: unpad the image
-    # raise NotImplementedError("TODO: compare the transformed images with the original ones")
+def _compare_extents(sdata0: SpatialData, sdata1: SpatialData):
+    e0 = get_extent(sdata0)
+    e1 = get_extent(sdata1, coordinate_system="transformed_back")
 
+    # we are not pixel perfect here, but "0.1-pixel perfect", I think this is due to an equivalent bug than then one
+    # that appears with rasterize (https://github.com/scverse/spatialdata/issues/165), probably originating from
+    # ndinterp(), which is used both by rasterize() and by transform()
+    assert all(np.allclose(e0[k], e1[k], atol=0.1) for k in set(e0.keys()).union(e1.keys()))
```
same as above

LucaMarconato replied:
Good point, actually I have renamed it to `are_extents_equal() -> bool` and moved it to `data_extent.py`

melonora reviewed (approved):
For record, the asserts were due to mypy complaining that a `BaseTransformation` does not have `keys`, while there is already an assert for the case that `get_all` is `True`. The other option other than the assert would be to ignore these mypy errors.

I think the PR can be approved, but I do think for future and with the refactor more comments can be added here and there to explain the code to make it more clear. Right now without the explanation on zoom it could be hard to follow or maybe that is just me:)

LucaMarconato commented on src/spatialdata/_core/data_extent.py:
```diff
@@ -289,9 +286,7 @@ def _(e: GeoDataFrame, coordinate_system: str = "global", exact: bool = True) ->
-    t = get_transformation(e, to_coordinate_system=coordinate_system)
-    assert isinstance(t, BaseTransformation)
-    transformed = transform(e, t)
+    transformed = transform(e, to_coordinate_system=coordinate_system)
```
Done.

LucaMarconato commented on src/spatialdata/_core/data_extent.py:
```diff
@@ -368,8 +359,10 @@ def _compute_extent_in_coordinate_system(
-    points = PointsModel.parse(df, coordinates={k: k for k in axes})
-    transformed_corners = pd.DataFrame(transform(points, transformation).compute())
+    d = get_transformation(element, get_all=True)
+    assert isinstance(d, dict)
```
you are right, I have removed it

LucaMarconato commented on src/spatialdata/_core/operations/_utils.py:
```diff
@@ -0,0 +1,117 @@
+    sdata_raster = SpatialData(images=dict(sdata.images), labels=dict(sdata.labels))
+    sdata_vector_transformed = sdata_vector.transform_to_coordinate_system(coordinate_system)
+
+    de = get_extent(sdata, coordinate_system=coordinate_system)
+    de_axes = tuple(de.keys())
```
Done, also for `de_axes`

LucaMarconato commented:
Thanks for the review. I have addressed all the comments (and left 3 conversations open to see if you agree with my proposal). I'll fix the docs, otherwise if you re-approve we are good to merge.

kevinyamauchi commented on src/spatialdata/_core/operations/transform.py:
```diff
@@ -104,24 +107,24 @@ def _transform_raster(
-    After an element has been transformed, this method is called to eventually prepend a particular transformation to
-    the existing transformations of the element. The transformation to prepend depends on the type of the element (
-    raster vs non-raster) and on the maintain_positioning flag.
+    """Adjust the transformations of an element after it has been transformed.
```
Please add a more clear docstring. I am not sure what "adjustments" are being made

LucaMarconato replied:
I renamed the function to  `_set_transformation_for_transformed_elements()` and fixed the docstring. In short, transformed elements don't have a transformation and this needs to be put inside. The transformation is not just the old transformation(s), but needs some "adjustments", which I now explain better.

kevinyamauchi commented on src/spatialdata/_core/operations/_utils.py:
```diff
@@ -0,0 +1,120 @@
+    target_depth: float | None = None,
+) -> SpatialData:
+    """
+    Transform the spatial data to match the data extent, keeping the positioning, and making all transformations equal.
```
Can this be another operation than scale? If not, I think we should say scale. I think "tranform" is ambiguous. 

Also what is meant by "making all transformations equal"?

LucaMarconato replied:
In `Transform the spatial data to match the data extent, keeping the positioning, and making all transformations equal.`

The `Transform` is referring to the application of the `transform()` operation. The transformation being applied can be anything, not just scale (affine, etc).

The returned object will be such that if we have the point coordinate (x, y), this will correspond to the pixel (y, x). The problem is that in this way we lose the positioning to the original data, so we still need to add scale and translations to make the returned data align with the original data. 

I will add these information to the docstring.

LucaMarconato replied:
I have also added an extra argument, called `maintain_positioning: bool = True`, which if false sets all the transformations to `Identity()`. I explained this argument in the docstring.

kevinyamauchi commented on tests/core/operations/test_transform.py:
```diff
@@ -459,14 +481,111 @@ def test_map_coordinate_systems_long_path(full_sdata):
+            remove_transformation(element, "global")
+
+    for element in full_sdata._gen_elements_values():
+        # from spatialdata.transformations.operations import _build_transformations_graph
+        #
+        # g = _build_transformations_graph(full_sdata)
+        # import networkx as nx
+        #
+        # nx.draw(g, with_labels=True)
+        # import matplotlib.pyplot as plt
+        #
+        # plt.show()
```
Do we need these commented lines still?

LucaMarconato replied:
removed.

kevinyamauchi commented on tests/core/operations/test_transform.py:
```diff
@@ -459,14 +481,111 @@ def test_map_coordinate_systems_long_path(full_sdata):
+            table=full_sdata.table,
+        )
+        temp["transformed_element"] = transformed_element
+        t = get_transformation_between_coordinate_systems(temp, temp["transformed_element"], "multi_hop_space")
+        a = t.to_affine_matrix(input_axes=("x",), output_axes=("x",))
+        d = get_transformation(transformed_element, get_all=True)
```
Can you please use more informative variable names?

LucaMarconato replied:
I agree, here it looks ugly. In general often in the code I use `t` for transformation, `a` for affine and `d` for the dict of transformations returned by `get_transformation(..., get_all=True)` (which is followed by `assert isinstance(d, dict)` to address `mypy`. So I think the `d` letter is ok.

kevinyamauchi reviewed (approved):
I think this looks good. I left some clarifying questions below. Please feel free to merge once you've addressed them.

melonora commented:
I also believe it can be merged after @kevinyamauchi comments are adressed.

LucaMarconato commented:
Thank you both for the review, I have addressed the review comments. 

I also added an argument `maintain_positioning: bool = True` to `transform_to_data_extent()`, which can be more intuitive for the user (I explained this in the docstring).

I also added some extra tests for checking better the results of `transform_to_data_extent()`, and test the new argument.

Merging now 🥳
